packages/itwinui-react/src/core/InputGrid/InputGrid.tsx

```diff
@@ -49,4 +70,207 @@ export const InputGrid = React.forwardRef((props, ref) => {
 
+/**
+ * Ensures that label, input and message are properly associated
+ * with each other, for accessibility purposes.
+ *
+ * - `Select` will be associated with label using `aria-labelledby`
+ * - Other inputs will be associated with label using `htmlFor`
+ * - Message will be associated with input/select using `aria-describedby`
+ */
+const useChildrenWithIds = (children: React.ReactNode) => {
+  const { labelId, inputId, messageId } = useSetup(children);
+
+  return React.useMemo(
+    () =>
+      React.Children.map(children, (child) => {
+        if (!React.isValidElement(child)) {
+          return child;
+        }
+
+        if (child.type === Label || child.type === 'label') {
+          return cloneElementWithRef(child, (child) => ({
+            htmlFor: inputId,
+            id: labelId,
+            ...child.props,
+          }));
+        }
+
+        if (child.type === StatusMessage) {
+          return cloneElementWithRef(child, (child) => ({
+            id: messageId,
+            ...child.props,
+          }));
+        }
+
+        if (
+          isInput(child) ||
+          child.type === InputWithDecorations ||
+          child.type === InputWithIcon
+        ) {
+          return handleCloningInputs(child, {
+            labelId,
+            inputId,
+            messageId,
+          });
+        }
+
+        return child;
+      }),
+    [children, inputId, labelId, messageId],
+  );
+};
+
+//-------------------------------------------------------------------------------
+
+/**
+ * Setup/prerequisite for `useChildrenWithIds` to gather information from children.
+ *
+ * Returns the following ids (prefers id from props, otherwise generates one)
+ *  - `labelId`
+ *  - `inputId`
+ *  - `messageId`
+ */
+const useSetup = (children: React.ReactNode) => {
+  const idPrefix = useId();
+
+  let labelId: string | undefined;
+  let inputId: string | undefined;
+  let messageId: string | undefined;
+
+  let hasLabel = false;
+  let hasSelect = false;
+
+  React.Children.forEach(children, (child) => {
+    if (!React.isValidElement(child)) {
+      return;
+    }
+
+    if (child.type === Label || child.type === 'label') {
+      hasLabel = true;
+      labelId = child.props.id || `${idPrefix}--label`;
+    }
+
+    if (child.type === StatusMessage) {
+      messageId = child.props.id || `${idPrefix}--message`;
+    }
+
+    const findInputId = (child: React.ReactNode) => {
+      if (!React.isValidElement(child)) {
+        return;
+      }
+      if (child.type === ComboBox) {
+        return child.props.inputProps?.id || `${idPrefix}--input`;
+      } else if (child.type !== Select) {
+        return child.props.id || `${idPrefix}--input`;
+      }
+    };
+
+    if (child.type === InputWithDecorations || child.type === InputWithIcon) {
+      React.Children.forEach(child.props.children, (child) => {
+        if (isInput(child)) {
+          inputId = findInputId(child);
```

Comment:
To be safe, should we stop the `forEach()` after we find the first input?

Reply:
i'm not sure about this; would there ever be more than one input?

there isn't a way to early exit from a `forEach` so i would need to use a `for..of` which requires converting the children to an array (i assume `Children.forEach` has some optimizations that i would be throwing away)

Reply:
actually I realized something:
since i'm no longer using state, it gets reset to `undefined` on every render anyway, so i can just use `||=` to only set it once on every render.

change made: [`0338ecb`](https://github.com/iTwin/iTwinUI/pull/1777/commits/0338ecbcfe395b6d27db731264d9b0cc92a47584)

Reply:
> i'm not sure about this; would there ever be more than one input?

I guess the officially supported case is with only one input, as demonstrated on our "Input with Decorations" docs page. So if more than one input is included for some reason, it's good that we're only considering the first one (after [`0338ecb` (#1777)](https://github.com/iTwin/iTwinUI/pull/1777/commits/0338ecbcfe395b6d27db731264d9b0cc92a47584) & https://github.com/iTwin/iTwinUI/pull/1777/commits/ed62b3e1ee8d4f98316c5f59ab2bf10bc479db3b)